de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SharedReferencePointMapper.java

```diff
@@ -0,0 +1,543 @@
+
+    public SharedReferencePointMapper() {
+        idToReferencePointMapping = new HashMap<String, IReferencePoint>();
+        referencePointToIDMapping = new HashMap<IReferencePoint, String>();
+        referencePointsOfUsers = new HashMap<User, List<String>>();
+        partiallySharedResourceMapping = new HashMap<IReferencePoint, Set<IResource>>();
+        completelySharedReferencePoints = new HashSet<IReferencePoint>();
+        partiallySharedReferencePoints = new HashSet<IReferencePoint>();
+    }
+
+    /**
+     * Adds a reference point to the set of currently shared reference points.
+     * <p>
+     * It is possible to "upgrade" a partially shared reference point to a
+     * completely shared reference point by just adding the same reference point
+     * with the same ID again that must now marked as not partially shared.
+     * </p>
+     * 
+     * @param id
+     *            the ID for the reference points
+     * @param referencePoint
+     *            the reference points to add
+     * @param isPartially
+     *            <code>true</code> if the reference point should be treated as
+     *            a partially shared reference point, <code>false</code> if it
+     *            should be treated as completely shared
+     * 
+     * @throws NullPointerException
+     *             if the ID or reference point is <code>null</code>
+     * @throws IllegalStateException
+     *             if the ID is already in use or the reference point was
+     *             already added
+     */
+    public synchronized void addReferencePoint(String id,
+        IReferencePoint referencePoint, boolean isPartially) {
+        boolean upgrade = false;
+
+        if (id == null)
+            throw new NullPointerException("ID is null");
+
+        if (referencePoint == null)
+            throw new NullPointerException("referencePoint is null");
+
+        String currentReferencePointID = referencePointToIDMapping
+            .get(referencePoint);
+        IReferencePoint currentReferencePoint = idToReferencePointMapping
+            .get(id);
+
+        if (currentReferencePointID != null
+            && !id.equals(currentReferencePointID)) {
+            throw new IllegalStateException("cannot assign ID " + id
+                + " to referencePoint " + referencePoint
+                + " because it is already registered with ID "
+                + currentReferencePointID);
+        }
+
+        if (currentReferencePoint != null
+            && !referencePoint.equals(currentReferencePoint)) {
+            throw new IllegalStateException("ID " + id + " for referencePoint "
+                + referencePoint + " is already used by referencePoint "
+                + currentReferencePoint);
+        }
+
+        if (isPartially
+            && partiallySharedReferencePoints.contains(referencePoint))
+            throw new IllegalStateException("referencePoint " + referencePoint
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Define a constant instead of duplicating this literal "referencePoint " 3 times. ([squid:S1192](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1192))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=257&issues=AWbeHvN3SgN4uCclRsGw&open=AWbeHvN3SgN4uCclRsGw)

[ISSUE_KEY]: AWbeHvN3SgN4uCclRsGw